File: java-learn/src/java_pj/Sudoku.java
package java_pj;

import java.util.Arrays;
import java.util.Scanner;

public class Sudoku {
	static final int TAM = 9;
	private static Scanner in = new Scanner(System.in);
	private static int[][] sudoku = new int[TAM][TAM];

	private static void limpiaVerificador(int[] verificador) {
		for (int i = 0; i < TAM; i++) {
			verificador[i] = 0;
		}

	}

	private static boolean sudoku_valido(int[][] puzzle) {
		// comprueba si es un sudoku valido
		int[] temp = { 0, 0, 0, 0, 0, 0, 0, 0, 0 }, temp2 = { 0, 0, 0, 0, 0, 0, 0, 0, 0 },
				temp3 = { 0, 0, 0, 0, 0, 0, 0, 0, 0 };
		for (int i = 0; i < TAM; i++) {
			for (int j = 0; j < TAM; j++) { // genera array con numero de repeticiones de un numero en una fila
				if (puzzle[i][j] != 0) {
					temp[puzzle[i][j] - 1]++;
				}
			}
			for (int j = 0; j < TAM; j++) { // verifica si se repite elemento en la fila
				if (temp[j] > 1) {
					return false;
				}
			}
			limpiaVerificador(temp);
			for (int j = 0; j < TAM; j++) { // genera array con numero de repeticiones de un numero en una columna
				if (puzzle[j][i] != 0) {
					temp[puzzle[j][i] - 1]++;
				}
			}
			for (int j = 0; j < TAM; j++) { // verifica si se repite elemento en la columna
				if (temp[j] > 1) {
					return false;
				}
			}
			limpiaVerificador(temp);
		}

		for (int i = 0; i < TAM; i++) {
			for (int k = 0; k < 3; k++) {
				if (puzzle[i][k] != 0) {
					temp[puzzle[i][k] - 1]++;
				}
				if (puzzle[i][k + 3] != 0) {
					temp2[puzzle[i][k + 3] - 1]++;
				}
				if (puzzle[i][k + 6] != 0) {
					temp3[puzzle[i][k + 6] - 1]++;
				}
				if ((1 + i) % 3 == 0) {
					for (int j = 0; j < TAM; j++) {
						if (temp[j] > 1) {
							return false;
						}
						if (temp2[j] > 1) {
							return false;
						}
						if (temp3[j] > 1) {
							return false;
						}
					}
					limpiaVerificador(temp);
					limpiaVerificador(temp2);
					limpiaVerificador(temp3);
				}
			}

		}
		return true;
	}

	private static boolean sudoku_acabado(int[][] puzzle) {
		// comprueba si un sudouku esta acabado
		for (int i = 0; i < TAM; i++) {
			for (int j = 0; j < puzzle[i].length; j++) {
				if (puzzle[i][j] == 0) {
					return false;
				}
			}
		}
		return true;
	}

	private static void nuevo_sudoku() {
		// crea un nuevo sudoku
		for (int i = 0; i < TAM; i++) {
			for (int j = 0; j < sudoku[i].length; j++) {
				if (sudoku[i][j] == 0) {

				}
			}
		}
	}

	public static void mostrar_sudoku() {
		// muestra el sudoku
		System.out.println(" -------------------------");
		for (int i = 0; i < TAM; i++) {
			System.out.print(" | ");
			for (int j = 0; j < sudoku[i].length; j++) {
				System.out.print(sudoku[i][j] + " ");
				if ((j + 1) % 3 == 0) {
					System.out.print("| ");
				}
			}
			System.out.println();
			if ((i + 1) % 3 == 0) {
				System.out.println(" -------------------------");
			}
		}

	}

	public static int[][] solucionar_sudoku(int[][] puzzle) {
		// da la opcion al usuario de completar el sudoku
		if (sudoku_acabado(puzzle) && sudoku_valido(puzzle)) {
			return puzzle;
		}
		for (int i = 0; i < TAM; i++) {
			for (int j = 0; j < TAM; j++) {
				if (puzzle[i][j] == 0) {
					for (int k = 1; k <= TAM; k++) {
						puzzle[i][j] = k;
						if (sudoku_valido(puzzle)) {
							int[][] res = solucionar_sudoku(puzzle);
							if (res != null) return res;
						}
						puzzle[i][j] = 0; // undo
					}
					return null;
				}
			}
		}
		return null;
	}

	public static void main(String[] args) {
		int[][] solvedSudoku = {
				{ 5, 3, 4, 6, 7, 8, 9, 1, 2 },
				{ 6, 7, 2, 1, 9, 5, 3, 4, 8 },
				{ 1, 9, 8, 3, 4, 2, 5, 6, 7 },
				{ 8, 5, 9, 7, 6, 1, 4, 2, 3 },
				{ 4, 2, 6, 8, 5, 3, 7, 9, 1 },
				{ 7, 1, 3, 9, 2, 4, 8, 5, 6 },
				{ 9, 6, 1, 5, 3, 7, 2, 8, 4 },
				{ 2, 8, 7, 4, 1, 9, 6, 3, 5 },
				{ 3, 4, 5, 2, 8, 6, 1, 7, 9 }
		};
		int[][] unsolvedSudoku = {
				{ 0, 0, 0, 0, 0, 0, 0, 0, 0 },
				{ 0, 0, 0, 1, 9, 5, 3, 4, 8 },
				{ 1, 9, 8, 3, 4, 2, 5, 6, 7 },
				{ 8, 5, 9, 7, 6, 1, 4, 2, 3 },
				{ 4, 2, 6, 8, 5, 3, 7, 9, 1 },
				{ 7, 1, 3, 9, 2, 4, 8, 5, 6 },
				{ 9, 6, 1, 5, 3, 7, 2, 8, 4 },
				{ 2, 8, 7, 4, 1, 9, 6, 3, 5 },
				{ 3, 4, 5, 2, 8, 6, 1, 7, 9 }
		};
		nuevo_sudoku();
		mostrar_sudoku();
		unsolvedSudoku = solucionar_sudoku(unsolvedSudoku);
		sudoku = unsolvedSudoku;
		mostrar_sudoku();
		if (sudoku_valido(solvedSudoku)) {
			System.out.println("valido");
		} else {
			System.out.println("no valido");
		}
		if (sudoku_valido(unsolvedSudoku)) {
			System.out.println("valido");
		} else {
			System.out.println("no valido");
		}
		if (Arrays.deepEquals(unsolvedSudoku,solvedSudoku)) {
			System.out.println("Funciona ole ole ole");
		} else {
			System.out.println("algo falla amigo");
		}
	}

}

File: static/js/sudoku.js
const TAM = 9;
const initial_vars = 31;
const new_sudoku = document.getElementById("nuevo");
const resolver = document.getElementById("resolver");
const comprobar = document.getElementById("comprobar");
let sudoku = Array(TAM).fill().map(() => Array(TAM).fill(0));

function limpiaVerificador(verificador) {
    for (let i = 0; i < TAM; i++) {
        verificador[i] = 0;
    }
}

function sudoku_valido(puzzle) {
    // comprueba si es un sudoku valido
    let temp = Array(9).fill(0);
    let temp2 = Array(9).fill(0);
    let temp3 = Array(9).fill(0);
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) { // genera array con numero de repeticiones de un numero en una fila
            if (puzzle[i][j] != 0) {
                temp[puzzle[i][j] - 1]++;
            }
        }
        for (let j = 0; j < TAM; j++) { // verifica si se repite elemento en la fila
            if (temp[j] > 1) {
                return false;
            }
        }
        limpiaVerificador(temp);
        for (let j = 0; j < TAM; j++) { // genera array con numero de repeticiones de un numero en una columna
            if (puzzle[j][i] != 0) {
                temp[puzzle[j][i] - 1]++;
            }
        }
        for (let j = 0; j < TAM; j++) { // verifica si se repite elemento en la columna
            if (temp[j] > 1) {
                return false;
            }
        }
        limpiaVerificador(temp);
    }

    for (let i = 0; i < TAM; i++) {
        for (let k = 0; k < 3; k++) {
            if (puzzle[i][k] != 0) {
                temp[puzzle[i][k] - 1]++;
            }
            if (puzzle[i][k + 3] != 0) {
                temp2[puzzle[i][k + 3] - 1]++;
            }
            if (puzzle[i][k + 6] != 0) {
                temp3[puzzle[i][k + 6] - 1]++;
            }
        }
        if ((1 + i) % 3 == 0) {
            for (let j = 0; j < TAM; j++) {
                if (temp[j] > 1) {
                    return false;
                }
                if (temp2[j] > 1) {
                    return false;
                }
                if (temp3[j] > 1) {
                    return false;
                }
            }
            limpiaVerificador(temp);
            limpiaVerificador(temp2);
            limpiaVerificador(temp3);
        }
    }

    return true;
}

function sudoku_acabado(puzzle) {
    // comprueba si un sudoku esta acabado (sobre tablero de numeros)
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) {
            if (puzzle[i][j] === 0) {
                return false;
            }
        }
    }
    return true;
}

function nuevo_sudoku() {
    // limpia
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) {
            sudoku[i][j] = 0;
        }
    }
}

function mostrar_sudoku() {
    // muestra el sudoku
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) {
            const cell = document.querySelector('.Sudoku-item[data-index="' + i + j + '"]');
            cell.value = sudoku[i][j] !== 0 ? sudoku[i][j] : "";
        }
    }
}

// Backtracking puro: usado como fallback cuando la propagacion de restricciones
// no es suficiente para resolver el puzzle completamente.
function solucionar_sudoku(puzzle) {
    if (sudoku_acabado(puzzle) && sudoku_valido(puzzle)) {
        return puzzle;
    }
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) {
            if (puzzle[i][j] === 0) {
                for (let k = 1; k <= TAM; k++) {
                    puzzle[i][j] = k;
                    if (sudoku_valido(puzzle)) {
                        const result = solucionar_sudoku(puzzle);
                        if (result !== null) return result;
                    }
                }
                puzzle[i][j] = 0; // deshacer
                return null;
            }
        }
    }
    return null;
}

// Convierte un tablero de Sets a un tablero de numeros.
// Las celdas resueltas (Set de 1 elemento) se convierten a su valor;
// las celdas sin resolver se dejan a 0 para que el backtracking las complete.
function sets_a_numeros(board) {
    return board.map(row =>
        row.map(cell => (cell.size === 1 ? [...cell][0] : 0))
    );
}

// Reduce las posibilidades de cada celda propagando las restricciones
// de filas, columnas y cajas 3x3.
function reduce_board(board, groups) {
    let changed = false;
    for (const group of groups) {
        // Recoge los valores ya fijos en el grupo
        const singles = new Set();
        for (const cell of group) {
            if (cell.size === 1) {
                singles.add([...cell][0]);
            }
        }
        // Elimina esos valores de las celdas aun no resueltas
        for (const cell of group) {
            if (cell.size > 1) {
                for (const num of singles) {
                    if (cell.has(num)) {
                        cell.delete(num);
                        changed = true;
                    }
                }
            }
        }
    }
    return changed;
}

// Solver principal: primero aplica propagacion de restricciones para reducir
// el espacio de busqueda y, si no basta, recurre al backtracking.
function solve_sudoku(puzzle) {
    // Construir tablero de Sets: celdas conocidas -> Set unitario, vacias -> {1..9}
    const board = puzzle.map(row =>
        row.map(val => (val !== 0 ? new Set([val]) : new Set([1, 2, 3, 4, 5, 6, 7, 8, 9])))
    );

    // Construir grupos: filas, columnas y cajas 3x3
    const groups = [];

    for (let i = 0; i < TAM; i++) {
        groups.push(board[i]); // fila i (array de Sets, no envuelto en otro array)
    }
    for (let j = 0; j < TAM; j++) {
        const col = [];
        for (let i = 0; i < TAM; i++) {
            col.push(board[i][j]);
        }
        groups.push(col);
    }
    for (let i = 0; i < TAM; i += 3) {
        for (let j = 0; j < TAM; j += 3) {
            const box = [];
            for (let k = 0; k < 3; k++) {
                for (let l = 0; l < 3; l++) {
                    box.push(board[i + k][j + l]);
                }
            }
            groups.push(box);
        }
    }

    // Propagar restricciones hasta que no haya cambios
    let changed = true;
    while (changed) {
        changed = reduce_board(board, groups);
    }

    // Convertir el tablero de Sets a numeros
    const numBoard = sets_a_numeros(board);

    // Si la propagacion ya resolvio el puzzle, devolver directamente
    if (sudoku_acabado(numBoard) && sudoku_valido(numBoard)) {
        return numBoard;
    }

    // Si no, usar backtracking sobre el tablero ya reducido (mucho menos trabajo)
    return solucionar_sudoku(numBoard);
}

function crea_sudoku() {
    nuevo_sudoku();
    for (let i = 0; i < initial_vars; i++) {
        let keep_on = true;
        while (keep_on) {
            const a = Math.floor(Math.random() * 9);
            const b = Math.floor(Math.random() * 9);
            if (sudoku[a][b] === 0) {
                for (let j = 1; j <= TAM; j++) {
                    sudoku[a][b] = j;
                    if (sudoku_valido(sudoku)) {
                        keep_on = false;
                        break;
                    }
                    sudoku[a][b] = 0;
                }
            }
        }
    }
    mostrar_sudoku();
}

function genera_sudoku() {
    let solucionable = false;
    while (!solucionable) {
        crea_sudoku();
        const sudoku2 = solve_sudoku(sudoku);
        if (sudoku2 !== null) {
            solucionable = true;
            sudoku = sudoku2;
        }
    }
}

function comprobar_sudoku() {
    for (let i = 0; i < TAM; i++) {
        for (let j = 0; j < TAM; j++) {
            const input = document.querySelector('.Sudoku-item[data-index="' + i + j + '"]');
            const td = document.querySelector('.Sudoku-td[data-index="' + i + j + '"]');
            const cls = parseInt(input.value) === sudoku[i][j] ? "acierto" : "error";
            td.classList.add(cls);
            setTimeout(() => td.classList.remove(cls), 2000);
        }
    }
}

genera_sudoku();
new_sudoku.addEventListener("click", () => genera_sudoku());
comprobar.addEventListener("click", () => comprobar_sudoku());
resolver.addEventListener("click", () => mostrar_sudoku());